ccl/importccl: TestExportImportBank failed

[(ccl/importccl).TestExportImportBank failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2764319&tab=buildLog) on [release-20.2@afc2dd516420ac5a93be8128d928be849ff15e4a](https://github.com/cockroachdb/cockroach/commits/afc2dd516420ac5a93be8128d928be849ff15e4a):

```
=== RUN   TestExportImportBank
--- FAIL: TestExportImportBank (25.09s)
    test_log_scope.go:158: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExportImportBank983234085
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestExportImportBank/null=
    --- FAIL: TestExportImportBank/null= (1.95s)
        exportcsv_test.go:98: error executing 'EXPORT INTO CSV $1 
            					WITH chunk_rows = $2, delimiter = '|' 
            					FROM SELECT * FROM bank AS OF SYSTEM TIME 1615409837986938323.0000000000': pq: failed to create file table to store uploaded file names: create-file-table: relation "userfiles_root_upload_files" already exists

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestExportImportBank PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #52807 ccl/importccl: TestExportImportBank failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestExportImportBank.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestExportImportBank failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2765045&tab=buildLog) on [release-20.2@dec0fec9bb9db92cee8007b6df228cf832fa683d](https://github.com/cockroachdb/cockroach/commits/dec0fec9bb9db92cee8007b6df228cf832fa683d):

```
=== RUN   TestExportImportBank
--- FAIL: TestExportImportBank (16.93s)
    test_log_scope.go:158: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExportImportBank954038030
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestExportImportBank/null=
    --- FAIL: TestExportImportBank/null= (10.26s)
        exportcsv_test.go:98: error executing 'EXPORT INTO CSV $1 
            					WITH chunk_rows = $2, delimiter = '|' 
            					FROM SELECT * FROM bank AS OF SYSTEM TIME 1615423264463691661.0000000000': pq: failed to create file table to store uploaded file names: create-file-table: relation "userfiles_root_upload_files" already exists

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestExportImportBank PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #52807 ccl/importccl: TestExportImportBank failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestExportImportBank.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestExportImportBank failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2765170&tab=buildLog) on [release-20.2@dec0fec9bb9db92cee8007b6df228cf832fa683d](https://github.com/cockroachdb/cockroach/commits/dec0fec9bb9db92cee8007b6df228cf832fa683d):

```
=== RUN   TestExportImportBank
--- FAIL: TestExportImportBank (21.65s)
    test_log_scope.go:158: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExportImportBank134943101
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestExportImportBank/null=
    --- FAIL: TestExportImportBank/null= (1.69s)
        exportcsv_test.go:98: error executing 'EXPORT INTO CSV $1 
            					WITH chunk_rows = $2, delimiter = '|' 
            					FROM SELECT * FROM bank AS OF SYSTEM TIME 1615427074927040129.0000000000': pq: failed to create file table to store uploaded file names: create-file-table: relation "userfiles_root_upload_files" already exists

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestExportImportBank PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #52807 ccl/importccl: TestExportImportBank failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestExportImportBank.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/importccl/exportcsv_test.go
+++ b/pkg/ccl/importccl/exportcsv_test.go
@@ -77,7 +77,7 @@ func TestExportImportBank(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	db, dir, cleanup := setupExportableBank(t, 3, 100)
+	db, _, cleanup := setupExportableBank(t, 3, 100)
 	defer cleanup()
 
 	// Add some unicode to prove FmtExport works as advertised.
@@ -85,32 +85,26 @@ func TestExportImportBank(t *testing.T) {
 	db.Exec(t, "UPDATE bank SET payload = NULL WHERE id % 2 = 0")
 
 	chunkSize := 13
+	baseExportDir := "userfile:///t/"
 	for _, null := range []string{"", "NULL"} {
 		nullAs := fmt.Sprintf(", nullas = '%s'", null)
 		nullIf := fmt.Sprintf(", nullif = '%s'", null)
 
 		t.Run("null="+null, func(t *testing.T) {
-			var files []string
+			exportDir := filepath.Join(baseExportDir, t.Name())
 
 			var asOf string
 			db.QueryRow(t, "SELECT cluster_logical_timestamp()").Scan(&asOf)
 
-			for _, row := range db.QueryStr(t,
-				fmt.Sprintf(`EXPORT INTO CSV 'nodelocal://0/t'
-					WITH chunk_rows = $1, delimiter = '|' %s
-					FROM SELECT * FROM bank AS OF SYSTEM TIME %s`, nullAs, asOf), chunkSize,
-			) {
-				files = append(files, row[0])
-				f, err := ioutil.ReadFile(filepath.Join(dir, "t", row[0]))
-				if err != nil {
-					t.Fatal(err)
-				}
-				t.Log(string(f))
-			}
+			db.Exec(t,
+				fmt.Sprintf(`EXPORT INTO CSV $1
+				WITH chunk_rows = $2, delimiter = '|' %s
+				FROM SELECT * FROM bank AS OF SYSTEM TIME %s`, nullAs, asOf), exportDir, chunkSize,
+			)
 
 			schema := bank.FromRows(1).Tables()[0].Schema
-			fileList := "'nodelocal://0/t/" + strings.Join(files, "', 'nodelocal://0/t/") + "'"
-			db.Exec(t, fmt.Sprintf(`IMPORT TABLE bank2 %s CSV DATA (%s) WITH delimiter = '|'%s`, schema, fileList, nullIf))
+			exportedFiles := filepath.Join(exportDir, "*")
+			db.Exec(t, fmt.Sprintf(`IMPORT TABLE bank2 %s CSV DATA ($1) WITH delimiter = '|'%s`, schema, nullIf), exportedFiles)
 
 			db.CheckQueryResults(t,
 				fmt.Sprintf(`SELECT * FROM bank AS OF SYSTEM TIME %s ORDER BY id`, asOf), db.QueryStr(t, `SELECT * FROM bank2 ORDER BY id`),

--- a/pkg/storage/cloudimpl/filetable/file_table_read_writer.go
+++ b/pkg/storage/cloudimpl/filetable/file_table_read_writer.go
@@ -268,7 +268,7 @@ func NewFileToTableSystem(
 		// TODO(adityamaru): Handle scenario where the user has already created
 		// tables with the same names not via the FileToTableSystem
 		// object. Not sure if we want to error out or work around it.
-		tablesExist, err := f.checkIfFileAndPayloadTableExist(ctx, e.ie)
+		tablesExist, err := f.checkIfFileAndPayloadTableExist(ctx, txn, e.ie)
 		if err != nil {
 			return err
 		}
@@ -779,7 +779,7 @@ func (f *FileToTableSystem) ReadFile(
 }
 
 func (f *FileToTableSystem) checkIfFileAndPayloadTableExist(
-	ctx context.Context, ie *sql.InternalExecutor,
+	ctx context.Context, txn *kv.Txn, ie *sql.InternalExecutor,
 ) (bool, error) {
 	tablePrefix, err := f.GetTableName()
 	if err != nil {
@@ -808,7 +808,7 @@ func (f *FileToTableSystem) checkIfFileAndPayloadTableExist(
 	tableExistenceQuery := fmt.Sprintf(
 		`SELECT table_name FROM [SHOW TABLES FROM %s] WHERE table_name=$1 OR table_name=$2`,
 		databaseSchema)
-	numRows, err := ie.ExecEx(ctx, "tables-exist", nil,
+	numRows, err := ie.ExecEx(ctx, "tables-exist", txn,
 		sessiondata.InternalExecutorOverride{User: security.RootUserName()},
 		tableExistenceQuery, fileTableName, payloadTableName)
 	if err != nil {

